contracts/upgradeable_contracts/amb_erc677_to_erc677/BasicAMBErc677ToErc677.sol

```diff
@@ -0,0 +1,188 @@
+pragma solidity 0.4.24;
+
+import "../../interfaces/IAMB.sol";
+import "../Ownable.sol";
+import "openzeppelin-solidity/contracts/AddressUtils.sol";
+import "../Initializable.sol";
+import "../ERC677Bridge.sol";
+import "../BaseOverdrawManagement.sol";
+import "../ReentrancyGuard.sol";
+import "../Upgradeable.sol";
+import "../Claimable.sol";
+import "../VersionableBridge.sol";
+
+contract BasicAMBErc677ToErc677 is
+    Initializable,
+    Ownable,
+    ReentrancyGuard,
+    Upgradeable,
+    Claimable,
+    VersionableBridge,
+    BaseOverdrawManagement,
+    ERC677Bridge
+{
+    bytes32 internal constant BRIDGE_CONTRACT = keccak256(abi.encodePacked("bridgeContract"));
+    bytes32 internal constant MEDIATOR_CONTRACT = keccak256(abi.encodePacked("mediatorContract"));
+    bytes32 internal constant REQUEST_GAS_LIMIT = keccak256(abi.encodePacked("requestGasLimit"));
+
+    function initialize(
+        address _bridgeContract,
+        address _mediatorContract,
+        address _erc677token,
+        uint256 _dailyLimit,
+        uint256 _maxPerTx,
+        uint256 _minPerTx,
+        uint256 _executionDailyLimit,
+        uint256 _executionMaxPerTx,
+        uint256 _requestGasLimit,
+        address _owner
+    ) external returns (bool) {
+        require(!isInitialized());
+        require(_minPerTx > 0 && _maxPerTx > _minPerTx && _dailyLimit > _maxPerTx);
+        require(_executionMaxPerTx < _executionDailyLimit);
+
+        uintStorage[DEPLOYED_AT_BLOCK] = block.number;
+        _setBridgeContract(_bridgeContract);
+        _setMediatorContract(_mediatorContract);
+        setErc677token(_erc677token);
+        uintStorage[DAILY_LIMIT] = _dailyLimit;
+        uintStorage[MAX_PER_TX] = _maxPerTx;
+        uintStorage[MIN_PER_TX] = _minPerTx;
+        uintStorage[EXECUTION_DAILY_LIMIT] = _executionDailyLimit;
+        uintStorage[EXECUTION_MAX_PER_TX] = _executionMaxPerTx;
+        _setRequestGasLimit(_requestGasLimit);
+        setOwner(_owner);
+        setInitialize();
+
+        emit DailyLimitChanged(_dailyLimit);
+        emit ExecutionDailyLimitChanged(_executionDailyLimit);
+
+        return isInitialized();
+    }
+
+    function fireEventOnTokenTransfer(address _from, uint256 _value) internal {
+        require(!lock());
+        bytes4 methodSelector = this.handleBridgedTokens.selector;
+        bytes memory data = abi.encodeWithSelector(methodSelector, _from, _value);
+        bridgeContract().requireToPassMessage(mediatorContract(), data, requestGasLimit());
+    }
+
+    function relayTokens(uint256 _value) external {
+        // This lock is to prevent calling fireEventOnTokenTransfer twice if a ERC677 token is used.
+        // When transferFrom is called, after the transfer, the ERC677 token will call onTokenTransfer from this contract
+        // which will call fireEventOnTokenTransfer.
+        require(!lock());
+        setLock(true);
```

Comment:
I am not sure that it will work properly:
1. `transferFrom` will call `callAfterTransfer`
2. it will call `onTokenTransfer` as so `contractFallback` is invoked
3. then `fireEventOnTokenTransfer` is called.
4. But the section is locked so `fireEventOnTokenTransfer` will revert.
5. It causes `contractFallback` to return `false` and `callAfterTransfer` reverts.
6. revert of `transferFrom` causes revert of `relayTokens`

Is this expected behavior?

Most probably you need to use `if (!lock())` without `revert()` in `fireEventOnTokenTransfer` (or `bridgeSpecificActionsOnTokenTransfer` as it is suggested above):
```solidity
function fireEventOnTokenTransfer(address _from, uint256 _value) internal {
    if (!lock()) {
        bytes4 methodSelector = this.handleBridgedTokens.selector;
        bytes memory data = abi.encodeWithSelector(methodSelector, _from, _value);
        bridgeContract().requireToPassMessage(mediatorContract(), data, requestGasLimit());
    }
}
```

Please also move ` setLock(true);` closer to the line with `token.transferFrom(from, to, _value);`:
```solidity
    setLock(true);
    token.transferFrom(from, to, _value);
    setLock(false);
```

Reply:
Good catch! and thanks for the suggestion. Please check https://github.com/poanetwork/poa-bridge-contracts/pull/273/commits/7ccacbf359528a2dd851d1dff03c9b5767d1df80